### PR TITLE
RSDK-4678 Tilt Compensation wit IMU

### PR DESCRIPTION
**Pull request by oliviamiller**
Adjusts the magnetometer readings when tilted -45 to 45 degrees. 

**Testing**
Screenshot below is with the imu being held at about 90 degrees and being tilted at various angles. As seen the compensated are more accurate but are still far off. Not sure how or if its possible to make this better? 
Screenshot 2023-09-27 at 2 59 38 PM

### CODE REVIEW

File: components/movementsensor/imuwit/imu.go [hunk 1]
```diff
@@ -124,26 +124,51 @@ func (imu *wit) getMagnetometer() (r3.Vector, error) {
...
-func calculateCompassHeading(x, y float64) float64 {
+// Helper function to calculate compass heading with tilt compensation.
+func (imu *wit) calculateCompassHeading() float64 {
+	pitch := imu.orientation.Pitch
+	roll := imu.orientation.Roll
+
+	var x, y float64
+
+	// Tilt compensation only works if the pitch and roll are between -45 and 45 degrees.
+	if roll >= -0.785 && roll <= 0.785 && pitch >= -0.785 && pitch <= 0.785 {
```
Comment (randhid) [inline components/movementsensor/imuwit/imu.go:147]: you can do `math.Abs(roll)` and `math.Abs(pitch)` less than 45deg if you'd like for fewer statements.  Maybe also make this a constant so t's read and not re-culcated (fortyfivedegsInRad to something better)

File: components/movementsensor/imuwit/imu.go [hunk 2]
```diff
@@ -124,26 +124,51 @@ func (imu *wit) getMagnetometer() (r3.Vector, error) {
...
-	return compass // change domain to 0 to 360
+
+	return compass
+}
+
+func (imu *wit) calculateTiltCompensation(roll, pitch float64) (float64, float64) {
+	// calculate adjusted magnetometer readings
+	xComp := imu.magnetometer.X*math.Cos(pitch) + imu.magnetometer.Z*math.Sin(pitch)
+	yComp := imu.magnetometer.X*math.Sin(roll)*math.Sin(pitch) +
+		imu.magnetometer.Y*math.Cos(roll) - imu.magnetometer.Z*math.Sin(roll)*math.Cos(pitch)
```
Comment (randhid) [inline components/movementsensor/imuwit/imu.go:169]: Can you add a comment on how accurate this is based don't he few tests you've tried?

**Review by randhid — APPROVED** [review]
Few nits! but looks good to me!

File: components/movementsensor/imuwit/imu.go [hunk 3]
```diff
@@ -124,26 +124,51 @@ func (imu *wit) getMagnetometer() (r3.Vector, error) {
...
+	}
+
 	// calculate -180 to 180 heading from radians
 	// North (y) is 0 so  the π/2 - atan2(y, x) identity is used
-	// directl
+	// directly
 	rad := math.Atan2(y, x) // -180 to 180 heading
 	compass := rutils.RadToDeg(rad)
 	compass = math.Mod(compass, 360)
 	compass = math.Mod(compass+360, 360)
```
Comment (susmitaSanyal) [inline components/movementsensor/imuwit/imu.go:160]: add a comment here about the range of the variable `compass`

**Review by susmitaSanyal — APPROVED** [review]
LGTM! just one nit.

**Pull request merged**